App/Http/Resolver.php

```diff
@@ -191,6 +192,64 @@ private function getRedirectUrl($url)
         return $response->getHeader('Location');
     }
 
+    /**
+     * Log the current rate limit remaining to the console.
+     *
+     * @param \GuzzleHttp\Message\Request $response
+     */
+    private function logRateLimitRemaining($response)
+    {
+        $rateLimitLimit = $response->getHeader('x-ratelimit-limit');
+        $rateLimitRemaining = $response->getHeader('x-ratelimit-remaining');
+
+        Utils::writeln(sprintf("Current rate limit remaining: {$rateLimitRemaining}/{$rateLimitLimit}"));
+    }
+
+    /**
+     * Checks and logs the rate limit, if the rate limit
+     * is exceeded, it will hold the script until the rate
+     * limit reset is over.
+     *
+     * @param string $url
+     */
+    private function handleRateLimit($url)
+    {
+        try {
+            $response = $this->client->get($url, [
```

Comment:
Does this request count for the rate-limiting logic? (The remaining quota will decrease?)

If so, we are losing download possibilies and also need to merge these requests:
```
$viemoUrl = $this->getRedirectUrl($downloadUrl);
$finalUrl = $this->getRedirectUrl($viemoUrl);
```

Reply:
Good point, its something that did cross my mind when first looking at that method, the url can get polled upwards of 3-4 times before the download initiates.

In my mind, its always the aim to reduce load on external sources. I didn't have time to look at it then, but I would have loved to refactor it to get it down to one request ideally.

The current patch, due to the current app structure, had to sit within its won request/method scope otherwise it would cloud the scope of the first method call. The first method call is also the one that, unfortunately, throws due to the rate limit being thrown on the first request.

Ideally I'd pass that first request down the execution chain and use it for any checks and downloading of the script so that we are only utilising one request, or as few as possible.

Sorry for a bit of incohesive waffle, hope it makes sense. TLDR; quite a big refactor of those two or three methods needs doing in order to solve it, something that I wanted to do but didn't have time, so we are only utilising one request/response, or as few as possible.

Reply:
No problem, you did the right thing 💪.

From what I can see, the refactor to only one download request shouldn't be that hard (if I'm not missing something):
```php
private function downloadLessonFromPath($html, $saveTo)
    {
        // ...
        try {
            $downloadUrl = Parser::getDownloadLink($html);
            
            // Save the response headers: Location (S3 url), Rate limit headers
            $response = $this->makeRequest($downloadUrl)

            // log info about the rate limiting (using the Rate Limiting headers)
            $this->handleRateLimit($response);

            // get the Location header value
            $viemoUrl = $this->getRedirectUrl($response);

            // get the Location header value
            $finalUrl = $this->getRedirectUrl($viemoUrl);
        } catch(NoDownloadLinkException $e) {
            // ...
            }
        }
        // ...
```

Reply:
If it works, go for it! 👍

Reply:
I don't have an active Laracasts account to test it (I'm not in the PHP world for a few years).